File: opti_coffee/solver/src/cpp/DistributionNetwork.cpp
#include "DistributionNetwork.h"
#include <iostream>

void DistributionNetwork::add_supplier(const Supplier& supplier) {
    suppliers.push_back(supplier);
}

void DistributionNetwork::add_roastery(const Roastery& roastery) {
    roasteries.push_back(roastery);
}

void DistributionNetwork::add_cafe(const Cafe& cafe) {
    cafes.push_back(cafe);
}

void DistributionNetwork::set_shipping_cost_from_supplier_to_roastery(const Supplier& supplier, const Roastery& roastery, double cost) {
    shipping_cost_from_supplier_to_roastery[std::make_pair(supplier.get_name(), roastery.get_name())] = cost;
}

void DistributionNetwork::set_shipping_cost_from_roastery_to_cafe(const Roastery& roastery, const Cafe& cafe, double cost) {
    shipping_cost_from_roastery_to_cafe[std::make_pair(roastery.get_name(), cafe.get_name())] = cost;
}

double DistributionNetwork::get_shipping_cost_from_supplier_to_roastery(const std::string& supplier_name, const std::string& roastery_name) const {
    auto it = shipping_cost_from_supplier_to_roastery.find(std::make_pair(supplier_name, roastery_name));
    return it != shipping_cost_from_supplier_to_roastery.end() ? it->second : std::numeric_limits<double>::infinity();
}

double DistributionNetwork::get_shipping_cost_from_roastery_to_cafe(const std::string& roastery_name, const std::string& cafe_name) const {
    auto it = shipping_cost_from_roastery_to_cafe.find(std::make_pair(roastery_name, cafe_name));
    return it != shipping_cost_from_roastery_to_cafe.end() ? it->second : std::numeric_limits<double>::infinity();
}

void DistributionNetwork::display_network_info() const {
    std::cout << "Suppliers:" << std::endl;
    for (const auto& supplier : suppliers) {
        std::cout << supplier.to_string() << std::endl;
    }
    std::cout << "\nRoasteries:" << std::endl;
    for (const auto& roastery : roasteries) {
        std::cout << roastery.to_string() << std::endl;
    }
    std::cout << "\nCafes:" << std::endl;
    for (const auto& cafe : cafes) {
        std::cout << cafe.to_string() << std::endl;
    }
}

File: opti_coffee/solver/src/cs/DistributionNetwork.cs
using System;
using System.Collections.Generic;

namespace CoffeeDistribution.Entities
{
    public class DistributionNetwork
    {
        public List<Supplier> Suppliers { get; private set; }
        public List<Roastery> Roasteries { get; private set; }
        public List<Cafe> Cafes { get; private set; }
        public Dictionary<(string, string), double> ShippingCostFromSupplierToRoastery { get; private set; }
        public Dictionary<(string, string), double> ShippingCostFromRoasteryToCafe { get; private set; }

        public DistributionNetwork()
        {
            Suppliers = new List<Supplier>();
            Roasteries = new List<Roastery>();
            Cafes = new List<Cafe>();
            ShippingCostFromSupplierToRoastery = new Dictionary<(string, string), double>();
            ShippingCostFromRoasteryToCafe = new Dictionary<(string, string), double>();
        }

        public void AddSupplier(Supplier supplier)
        {
            Suppliers.Add(supplier);
        }

        public void AddRoastery(Roastery roastery)
        {
            Roasteries.Add(roastery);
        }

        public void AddCafe(Cafe cafe)
        {
            Cafes.Add(cafe);
        }

        public void SetShippingCostFromSupplierToRoastery(Supplier supplier, Roastery roastery, double cost)
        {
            ShippingCostFromSupplierToRoastery[(supplier.Name, roastery.Name)] = cost;
        }

        public void SetShippingCostFromRoasteryToCafe(Roastery roastery, Cafe cafe, double cost)
        {
            ShippingCostFromRoasteryToCafe[(roastery.Name, cafe.Name)] = cost;
        }

        public double GetShippingCostFromSupplierToRoastery(string supplierName, string roasteryName)
        {
            return ShippingCostFromSupplierToRoastery.TryGetValue((supplierName, roasteryName), out double cost) ? cost : double.PositiveInfinity;
        }

        public double GetShippingCostFromRoasteryToCafe(string roasteryName, string cafeName)
        {
            return ShippingCostFromRoasteryToCafe.TryGetValue((roasteryName, cafeName), out double cost) ? cost : double.PositiveInfinity;
        }

        public void DisplayNetworkInfo()
        {
            Console.WriteLine("Suppliers:");
            foreach (var supplier in Suppliers)
            {
                Console.WriteLine(supplier);
            }

            Console.WriteLine("\nRoasteries:");
            foreach (var roastery in Roasteries)
            {
                Console.WriteLine(roastery);
            }

            Console.WriteLine("\nCafes:");
            foreach (var cafe in Cafes)
            {
                Console.WriteLine(cafe);
            }
        }
    }
}